sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -434,18 +511,139 @@ case class SortMergeJoinExec(
+          createJoinKey(ctx, leftRow, leftUpperKeys, left.output),
+          createJoinKey(ctx, rightRow, rightLowerKeys, right.output),
+          createJoinKey(ctx, rightRow, rightUpperKeys, right.output))
+      }
+      else {
+        (Nil, Nil, Nil, Nil)
+      }
+
+    val secRangeDataType = if (leftLowerKeys.size > 0) { leftLowerKeys(0).dataType }
+      else if (leftUpperKeys.size > 0) { leftUpperKeys(0).dataType }
+      else null
+    val secRangeInitValue = CodeGenerator.defaultValue(secRangeDataType)
+
+    val (leftLowerSecRangeKey, leftUpperSecRangeKey, rightLowerSecRangeKey, rightUpperSecRangeKey) =
+      if (useInnerRange) {
+        (ctx.addBufferedState(secRangeDataType, "leftLowerSecRangeKey", secRangeInitValue),
+          ctx.addBufferedState(secRangeDataType, "leftUpperSecRangeKey", secRangeInitValue),
+          ctx.addBufferedState(secRangeDataType, "rightLowerSecRangeKey", secRangeInitValue),
+          ctx.addBufferedState(secRangeDataType, "rightUpperSecRangeKey", secRangeInitValue))
+      }
+      else {
+        (null, null, null, null)
+      }
+
+    // A queue to hold all matched rows from right side.
+    val clsName = if (useInnerRange) classOf[InMemoryUnsafeRowQueue].getName
+      else classOf[ExternalAppendOnlyUnsafeRowArray].getName
 
     val spillThreshold = getSpillThreshold
     val inMemoryThreshold = getInMemoryThreshold
 
-    // Inline mutable state since not many join operations in a task
     val matches = ctx.addMutableState(clsName, "matches",
       v => s"$v = new $clsName($inMemoryThreshold, $spillThreshold);", forceInline = true)
-    // Copy the left keys as class members so they could be used in next function call.
     val matchedKeyVars = copyKeys(ctx, leftKeyVars)
 
+    val lowerCompop = lowerSecondaryRangeExpression.map {
+      case GreaterThanOrEqual(_, _) => "<"
+      case GreaterThan(_, _) => "<="
+      case _ => ""
+    }.getOrElse("")
+    val upperCompop = upperSecondaryRangeExpression.map {
+      case LessThanOrEqual(_, _) => ">"
+      case LessThan(_, _) => ">="
+      case _ => ""
+    }.getOrElse("")
+    val lowerCompExp = if (!useInnerRange || lowerSecondaryRangeExpression.isEmpty) ""
+      else s" || (comp == 0 && ${leftLowerSecRangeKey.value} " +
+        s"$lowerCompop ${rightLowerSecRangeKey.value})"
+    val upperCompExp = if (!useInnerRange || upperSecondaryRangeExpression.isEmpty) ""
+      else s" || (comp == 0 && ${leftUpperSecRangeKey.value} " +
+        s"$upperCompop ${rightUpperSecRangeKey.value})"
+
+    logDebug(s"lowerCompExp: $lowerCompExp")
+    logDebug(s"upperCompExp: $upperCompExp")
+
+    // Add secondary range dequeue method
+    if (!useInnerRange || lowerSecondaryRangeExpression.isEmpty ||
+        rightLowerKeys.size == 0 || rightUpperKeys.size == 0) {
+      ctx.addNewFunction("dequeueUntilUpperConditionHolds",
+        "private void dequeueUntilUpperConditionHolds() { }",
+        inlineToOuterClass = true)
```

Comment:
do we really need this? can't we just use the returned name?

Reply:
can we avoid to use `inlineToOuterClass = true`? I think we can avoid doing that.

Reply:
Ah, OK. I'll investigate.